compiler/AST/foralls.cpp

```diff
@@ -230,3 +234,763 @@ bool astUnderFI(const Expr* ast, ForallIntents* fi) {
+// It would be fn->retType, alas protoIteratorClass() messes with that.
+// This helper is ForallStmt-specific because of the error message it issues.
+//
+static QualifiedType fsIterYieldType(FnSymbol* iterFn,
+                                     CallExpr* iterCall,
+                                     bool gotSA)
+{
+  if (iterFn->hasFlag(FLAG_ITERATOR_FN)) {
+    IteratorInfo* ii = iterFn->iteratorInfo;
+    INT_ASSERT(ii);
+    return ii->getValue->getReturnQualType();
+  } else {
+    // e.g. "proc these() return _value.these();"
+    AggregateType* retType = toAggregateType(iterFn->retType);
+    if (retType && retType->symbol->hasFlag(FLAG_ITERATOR_RECORD)) {
+      FnSymbol* iterator = retType->iteratorInfo->iterator;
+      INT_ASSERT(iterator->hasFlag(FLAG_ITERATOR_FN)); // no more recursion?
+      return fsIterYieldType(iterator, iterCall, gotSA);
+    } else {
+      USR_FATAL_CONT(iterCall, "The iterable-expression resolves to a non-iterator function '%s' when looking for %s iterator", iterFn->name, gotSA ? "standalone" : "leader");
+      USR_PRINT(iterFn, "The function '%s' is declared here", iterFn->name);
+      USR_STOP();
+      return (QualifiedType)0; // dummy
+    }
+  }
+}
+
+// Like in build.cpp, here for ForallStmt.
+static BlockStmt*
+buildFollowLoop(VarSymbol* iter,
+                VarSymbol* leadIdxCopy,
+                VarSymbol* followIter,
+                VarSymbol* followIdx,
+                BlockStmt* loopBody,
+                Expr*      ref,
+                bool       fast,
+                bool       zippered) {
+  BlockStmt* followBlock = new BlockStmt();
+  ForLoop*   followBody  = new ForLoop(followIdx, followIter, loopBody, zippered);
+
+  // not needed:
+  //destructureIndices(followBody, indices, new SymExpr(followIdx), false);
+
+  followBlock->insertAtTail(new DefExpr(followIter));
+
+  if (fast) {
+
+    if (zippered) {
+      followBlock->insertAtTail("'move'(%S, _getIteratorZip(_toFastFollowerZip(%S, %S)))", followIter, iter, leadIdxCopy);
+    } else {
+      followBlock->insertAtTail("'move'(%S, _getIterator(_toFastFollower(%S, %S)))",       followIter, iter, leadIdxCopy);
+    }
+  } else {
+
+    if (zippered) {
+      followBlock->insertAtTail("'move'(%S, _getIteratorZip(_toFollowerZip(%S, %S)))",     followIter, iter, leadIdxCopy);
+    } else {
+      followBlock->insertAtTail("'move'(%S, _getIterator(_toFollower(%S, %S)))",           followIter, iter, leadIdxCopy);
+    }
+  }
+
+  ref->insertAfter(followBlock); // otherwise it wouldn't normalize
+  normalize(followBlock);
+  followBlock->remove();
+
+  followBlock->insertAtTail(new DefExpr(followIdx));
+
+  followBlock->insertAtTail("{TYPE 'move'(%S, iteratorIndex(%S)) }", followIdx, followIter);
+
+  followBlock->insertAtTail(followBody);
+  followBlock->insertAtTail(new CallExpr("_freeIterator", followIter));
+
+  return followBlock;
+}
+
+static SymExpr* normalizeFSIterExpr(ForallStmt* pfs, SymExpr* origSE)
```

Comment:
I'd prefer not to call this "normalize" since that word means something else in our compiler.
Also, it appears that this is only called from buildForallParIterCall and that is called from resolveParallelIteratorAndForallIntents. These last two can't possibly have reasonable method names, because we have to redirect to _value for for loops in addition to forall loops over _domain/_array. (Or else I'm missing something).